Fix #10532

iamajoe opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10553)

<!-- Reviewable:end -->

Commit:
[fix] Input line height and styling

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @pcwalton (or someone else) soon.

pcwalton commented:
Where does that line height value come from?

Sendoushi commented:
Comparison with Firefox. Photoshop to check if everything is as supposed.

perlun commented on resources/servo.css:
```diff
@@ -1,5 +1,5 @@
-input                   { background: white; min-height: 1.0em; padding: 0em; padding-left: 0.25em; padding-right: 0.25em; border: solid lightgrey 1px; color: black; white-space: nowrap; text-align: left; }
-textarea                { background: white; min-height: 1.0em; padding: 0em; padding-left: 0.25em; padding-right: 0.25em; border: solid lightgrey 1px; color: black; white-space: pre; }
+input                   { background: white; min-height: 1.0em; padding: 0em; padding-left: 0.25em; padding-right: 0.25em; border: solid lightgrey 1px; color: black; font-family: sans-serif; font-size: 0.7em; white-space: nowrap; text-align: left; line-height: 1.8; }
+textarea                { background: white; min-height: 1.0em; padding: 0em; padding-left: 0.25em; padding-right: 0.25em; border: solid lightgrey 1px; color: black; font-family: sans-serif; font-size: 0.7em; white-space: pre; }
```
Shouldn't we try to split these properties onto multiple lines? It's awfully hard to meaningfully see the changes in the diff when the style is structured like that.

(I know it was the existing structure; I am just suggesting that we should clean it up. :smile:)

Sendoushi replied:
I totally agree. I also find it weird to work with these.
Also... This file will definitely grow so, besides an import system we should maybe start commenting with sections? For example `/* Form */`.

pcwalton commented:
Can you file a follow-up bug to get this added to the spec? https://html.spec.whatwg.org/multipage/rendering.html#form-controls

Once that's done, I'm fine with this patch.

aethanyc commented:
Firefox use `line-height: normal` in [forms.css](https://dxr.mozilla.org/mozilla-central/rev/21bf1af375c1fa8565ae3bb2e89bd1a0809363d4/layout/style/res/forms.css#91l). 

Do we need a reftest to check these two pages are the same? These two look the same on Firefox. I'm not sure this is in the scope of this bug because even if we set the `line-height` of the `input` element, the line containing the `input` still have the height from the parent.
```html
<div style="line-height: 100px;">
  <input placeholder="foobar">
</div>
```
```html
<div>
  <input placeholder="foobar">
</div>
```

Sendoushi commented:
@aethanyc Yes I noticed that `line-height: normal` in the firefox code before. The problem is that if I set with normal it won't get the right size. Setting line-height as I did, reseted to the right size but... maybe there is other way to set the default on servo so we could use that normal?

@pcwalton What would you want me to write? I'm not used to write these things :P New to contributing but want to keep on doing it!

paulrouget commented:
(just a unrelated note: please don't name your PR "Fix #xxxx". It makes it a bit difficult to triage PRs and go through github notifications. And thank you for the PR ^^)

Sendoushi commented:
@paulrouget Sure. How would you prefer me to name the PR? So that next time I've that in attention?

aethanyc commented:
@Sendoushi I'm new to servo, too. Let's wait servo experts to answer your question.

KiChjang commented:
@Sendoushi It'd be nice if you can have a descriptive title of what the PR is about. Coincidentally, your commit message for this PR is good enough as a PR title.

jdm commented:
https://html.spec.whatwg.org/multipage/rendering.html#the-input-element-as-a-text-entry-widget is probably a more appropriate place in the specification to define this. The issue should be filed at https://github.com/whatwg/html/issues .

perlun commented:
@jdm, @pcwalton - I think @Sendoushi would need some help here in what to write in the whatwg issue. Could either one of you (who know what to write :smile:) report the issue there and refer back to this PR?

Sendoushi commented:
Yes please. I have no idea on the next steps...

jdm commented:
I have filed https://github.com/whatwg/html/issues/1046 as requested.

pcwalton commented:
Given that upstream spec states that this could reasonably be covered under the magic "binding" property, I think it's safe to say the spec allows us to do this today, so r=me.

pcwalton commented:
Oh wait... r=me with a reftest added.

Commit:
Merge branch 'master' of https://github.com/servo/servo

Ms2ger commented:
(The spec problem is not so much "are we allowed to this" but "should we be *required* to do this")

Commit:
[new] Input line height ref test

Commit:
Merge branch 'master' of https://github.com/servo/servo into input-tests

Sendoushi commented:
Here is the reftest: https://github.com/servo/servo/pull/10648 is it ok?

notriddle commented:
There's no need to merge with master, or to have a separate pull request.

Ms2ger committed:
Update js.

jdm committed:
Use DOM width/height for canvas display list item.

Commit:
[fix] Test for input in manifest

Commit:
Merge branch 'master' of https://github.com/servo/servo into input-tests

Sendoushi commented:
This branch now has the tests may you please confirm that everything is as expected?

jdm committed:
Ensure that we aren't missing any errors in the tidy self-test.

Commit:
Remove modelines everywhere but tests/wpt

perlun committed:
Improve properties.mako.rs file structure, take 2

This is a new attempt of #10586, after Simon Sapin's great cleanups in #10749 has landed. I have adjusted the changes to the new structure that was introduced, and also only done a few of the longhand ones. Will certainly continue on this as soon as we have a basic agreement that this style is reasonable.

notriddle commented:
You still have a bunch of merge commits in here (if you look at the commits tab, you can see all three of them). You can use git's rebase command to remove them.

Sendoushi commented:
Oh. All in one commit. On it!

Ms2ger committed:
Update web-platform-tests to revision 0a518aaff73532a26e175789f7e75fa99593ac64

Commit:
Merge branch 'master' of https://github.com/servo/servo

Sendoushi commented:
Being hard to rebase this. Can anyone help?

notriddle commented:
[Here's a nice article on how to do it, better than one I'd write probably](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html)

SimonSapin committed:
Blind attempt at fixing a Python import error on Windows.

See #10789.

mbrubeck committed:
Disable /webgl/conformance-1.0.3/conformance/more/conformance/quickCheckAPI-S_V.html

Because Servo only sometimes crashes on its random data (#10698).

jdm committed:
Disable frequent intermittently-passing tests.

perlun committed:
Applied a bunch of suggestions from Simon.

perlun commented:
@Sendoushi - in this case you _should_ be able to do something like:

```
git fetch --all # Gives you the latest changes from all remotes
git rebase origin/master # This will replay your changes on top of current master, removing merge commits automatically
git push -f # Will force-push the branch to your fork
```

It shouldn't be harder than that. Give it a try and see.

Ms2ger committed:
Disable WebGL tests.

They're not stable enough right now, with no expectation of improvement
any time soon.

Sendoushi commented:
@notriddle checked that but it is hard to spot my commits now.
Tried that @perlun now it is even worse :s

perlun commented:
What does `git remote -v` show? I think the problem may be that `origin` isn't the `servo` remote in this case, but your local fork, so you didn't rebase on top of what I expected.

Also, `git rebase -i origin/master` may be an idea where you can explicitly remove the lines for the commits that you don't want to include in the feature branch. Give it a try and see if you can make it work.

Sendoushi commented:
@perlun yes, servo remote is "upstream". I tried -i but the problem with that is that i already have a lot of commits which aren't my own and it is pretty hard for me to go through each commit. I guess at this point I should just re-fork and re-do the whole thing.

perlun commented:
Try 'git rebase upstream/master' and then force push. Should do the trick, no?

Sendoushi commented:
That makes sense. Pulled upstream/master (everything up to date), rebase upstream/master it and conflict errors on `test-helpers.sub.js` and `MANIFEST.json`

KiChjang commented:
Something is horribly wrong, not only do I see multiple commits, I see commits that are not from the PR author.

perlun commented:
Weird. I'll see if I can sort it out.

perlun commented:
@Sendoushi - I think I got the branch fixed. The steps I took:

1. Added your repo as a remote in my own working copy: `git remote add sendoushi https://github.com/Sendoushi/servo.git`
1. Fetched form all remotes, to ensure I had your changes and also the latest changes from servo upstream (the remote named `origin` in my case).
1. Checked out your branch and gave it a sensible name: `checkout -b sendoushi-master sendoushi/master`
1. Interactively rebased, to see the list of commits: `git rebase -i origin/master`. Gave me this output and no other commits:

```
pick 4cd043e [fix] Input line height and styling
pick db9244d [new] Input line height ref test
pick eff4835 [fix] Test for input in manifest
pick 27eea7b Improve properties.mako.rs file structure, take 2
```

The last one is a change I've made; I don't know how it made its way there. :smile: It's already merged into servo upstream, but I kept it anyway and let the rebase start.

I got an error when it attempted to apply that commit. Since I knew it's already merged and doesn't belong to your branch anyway, I skipped it in the rebase: `git rebase --skip`

The rebase was done. I pushed the result to my own fork: `git push perlun sendoushi-master`

And then I created the PR. Since I rebased your commits, they still have you as the author and you should get the credit you (rightfully) deserve.

Hope this helped! :smile:

jdm commented:
Closing in favour of #10809. Thanks @perlun!

Sendoushi commented:
@perlun it helped :) rebase is still a mysterious wizardry for me

perlun commented:
@Sendoushi - it's one of the hardest parts of git perhaps, but don't be afraid: you can learn it as well. I happen to use it more or less every day in my work so I'm very used to the workflow.

(Think of rebase as "replay the commits from my current branch on top of some other commit than the commit I branched out from. Problems can occur if you rebase on top of the wrong commit etc, but once you learn it, it's one of the things with git that you tend to like the most.)
